protocols/pop3/src/main/java/org/apache/james/protocols/pop3/core/POP3MessageCommandDelegate.java

```diff
@@ -0,0 +1,101 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.protocols.pop3.core;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.james.protocols.api.ProtocolSession;
+import org.apache.james.protocols.api.Request;
+import org.apache.james.protocols.api.Response;
+import org.apache.james.protocols.pop3.POP3Response;
+import org.apache.james.protocols.pop3.POP3Session;
+import org.apache.james.protocols.pop3.mailbox.MessageMetaData;
+
+import com.google.common.collect.ImmutableList;
+
+public abstract class POP3MessageCommandDelegate {
+
+    private final String command;
+    
+    public POP3MessageCommandDelegate(Collection<String> commands) {
+        if (commands.isEmpty()) {
+            throw new IllegalArgumentException("must provide at least one command keyword");
+        }
+        this.command = commands.iterator().next();
+    }
+
+    public Response handleMessageRequest(POP3Session session, Request request) {
+        if (session.getHandlerState() == POP3Session.TRANSACTION) {
+
+            String parameters = request.getArgument();
+            if (parameters == null) {
+                return handleSyntaxError();
+            }
+
+            final List<Integer> args;
+            try {
+                args = Arrays.stream(parameters.trim().split(" "))
+                    .map(String::trim).filter(arg -> !arg.isEmpty()).map(Integer::parseInt)
+                    .collect(Collectors.toUnmodifiableList());
+            } catch (NumberFormatException nfe) {
+                return handleSyntaxError();
+            }
+            if (args.isEmpty()) {
+                return handleSyntaxError();
+            }
+            int num = args.get(0);
+
+            try {
+
+                MessageMetaData data = getMetaData(session, num);
+
+                List<String> deletedUidList = session.getAttachment(POP3Session.DELETED_UID_LIST, ProtocolSession.State.Transaction).orElse(ImmutableList.of());
+                if (!deletedUidList.contains(data.getUid())) {
+                    return handleMessageExists(session, data, args);
+                } else {
+                    return new POP3Response(POP3Response.ERR_RESPONSE, "Message (" + num + ") already deleted.");
+                }
+
+            } catch (IOException ioe) {
+                return new POP3Response(POP3Response.ERR_RESPONSE, "Message (" + num + ") does not exist.");
+            }
+        } else {
+            return POP3Response.ERR;
```

Comment:
This is a 40+ lines of codes method, we should split it in smaller chunks.

Here we can distinguish the following responsibilities:
 - Parsing (we could introduce a small POJO to combine num and args in the return type)
 - Processing

Reply:
OK, now I see what you meant with the parsing POJO. Done in latest commit.